src/addspot.c

```diff
@@ -61,28 +63,119 @@ void add_to_spots(char *call, freq_t freq) {
     lanspotflg = false;
 }
 
+static freq_t ask_frequency() {
+    char frequency[16];
+
+    attron(COLOR_PAIR(C_LOG) | A_STANDOUT);
+
+    mvaddstr(13, 20, "freq.: ");
+    echo();
+    getnstr(frequency, 7);
+    noecho();
+    return atof(frequency) * 1000.0;
+}
 
 void addspot(void) {
 
-    char frequency[8];
 
     if (strlen(hiscall) < 3) {
 	return;
     }
 
     if (!trx_control) {
-
-	attron(COLOR_PAIR(C_LOG) | A_STANDOUT);
-
-	mvaddstr(13, 20, "freq.: ");
-	echo();
-	getnstr(frequency, 7);
-	noecho();
-	freq = atof(frequency) * 1000.0;
+	freq = ask_frequency();
     }
 
     add_to_spots(hiscall, freq);
 
     hiscall[0] = '\0';
+}
+
+/* send spot to cluster */
+
+#define MAX_SPOT_AGE 120
+
+/* find last qso record in qso_array, return NULL if no one found */
+static struct qso_t *find_last_qso() {
+    int i = nr_qsos;
+    while (i > 0) {
+	struct qso_t *last_qso = g_ptr_array_index(qso_array, nr_qsos-1);
+	if (!last_qso->is_comment) {
+	    return last_qso;
+	}
+	i--;
+    }
+    return NULL;
+}
 
+static bool spot_to_old(const struct qso_t *qso) {
+    return ((get_time() - qso->timestamp) > MAX_SPOT_AGE);
+}
+
+/* get call and frequency of spot,
+ * try first to use hiscall as spot call. If empty use data from last qso if
+ * not to old.
+ * Returns true if valid spot data found.
+ */
+bool get_spot_data(char **spot_call, freq_t *spot_freq) {
+    if (strlen(hiscall) > 2) {
+	*spot_call = g_strdup(hiscall);
+	if (trx_control) {
+	    *spot_freq = freq/1000.;
+	} else {
+	    *spot_freq = ask_frequency();
+	}
+    } else {
+	struct qso_t *last_qso = find_last_qso();
+	if (last_qso == NULL) {
+	    return false;
+	}
+	if (spot_to_old(last_qso)) {
+	    return false;
+	}
+	*spot_call = g_strdup(last_qso->call);
+	if (last_qso->freq < 1.) {	/* not set */
+	    *spot_freq = ask_frequency();
+	} else {
+	    *spot_freq = last_qso->freq;
+	}
+    }
+    return true;
+}
+
+gchar *prepare_spot(void) {
+    gchar *spot_line;
+    gchar *spot_call;
+    freq_t spot_freq;
+    if (get_spot_data(&spot_call, &spot_freq)) {
+	spot_line = g_strdup_printf("DX %.1f %s ", spot_freq,
+		spot_call);
+	g_free(spot_call);
+    } else {
+	beep();
+	spot_line = g_strdup("");
+    }
+    return spot_line;
+}
+
+
+void send_spot(void) {
+    gchar *spot_line = prepare_spot();
+    int c;
+
+    if (strlen(spot_line) > 0) {
+
+	clear_line(LINES - 1);
+	mvprintw(LINES - 1, 0, "> %s", spot_line);
+	refreshp();
+	move(LINES - 1, strlen(spot_line) + 2);
+	while ((c = key_get()) != ESCAPE && c != '\n' && c != KEY_ENTER)
+	    ;
+	if (c != ESCAPE) {
+	    gchar *line = g_strconcat(spot_line, "\n", NULL);
```

Comment:
'send_cluster' will be dropped later. That is not the problem.

But the code from 'send_to_cluster' is duplicated around line 1080 of 'splitscreen.c' which should 
be refactored using 'send_to_cluster'. Unluckily the added '\n' is sent to 'addtest(line)' before. If we move adding '\n' to 'send_to_cluster()' it would be missing in the string sent to 'addtext()'. That function is quite nasty and I am not sure I want to touch it at the moment.

Reply:
yes, `splitscreen` also deserves a re-work sometime later.